accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,511 @@
+    /// e.g. the request duration.
+    /// </summary>
+    public class Histogram<T> : Instrument<T> where T : unmanaged
+    {
+        public void Record(T measurement) { throw null; }
+        public void Record(
+                        T measurement,
+                        KeyValuePair<string, object?> tag1) { throw null; }
+        public void Record(
+                        T measurement,
+                        KeyValuePair<string, object?> tag1,
+                        KeyValuePair<string, object?> tag2) { throw null; }
+        public void Record(
+                        T measurement,
+                        KeyValuePair<string, object?> tag1,
+                        KeyValuePair<string, object?> tag2,
+                        KeyValuePair<string, object?> tag3) { throw null; }
+        public void Record(T measurement,
+                            ReadOnlySpan<<string, object?>> tags) { throw null; }
+        public void Record(
+                        T measurement,
+                        params KeyValuePair<string, object?>[] tags) { throw null; }
+    }
+}
+
+```
+
+### Observable Instruments
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// ObservableCounter is an observable Instrument that reports monotonically increasing value(s)
+    /// when the instrument is being observed.
+    /// e.g. CPU time (for different processes, threads, user mode or kernel mode).
+    /// </summary>
+    public class ObservableCounter<T> : ObservableInstrument<T> where T : unmanaged
+    {
+        protected override IEnumerable<Measurement<T>> Observe() { throw null; }
+    }
+
+    /// <summary>
+    /// ObservableUpDownCounter is an observable Instrument that reports additive value(s)
+    /// when the instrument is being observed.
+    /// e.g. the process heap size
+    /// </summary>
+    public class ObservableUpDownCounter<T> : ObservableInstrument<T> where T : unmanaged
+    {
+        protected override IEnumerable<Measurement<T>> Observe() { throw null; }
+    }
+
+    /// <summary>
+    /// ObservableGauge is an observable Instrument that reports non-additive value(s)
+    /// when the instrument is being observed.
+    /// e.g. the current room temperature
+    /// </summary>
+    public class ObservableGauge<T> : ObservableInstrument<T> where T : unmanaged
+    {
+        protected override IEnumerable<Measurement<T>> Observe() { throw null; }
+    }
+}
+```
+
+### MeterListener
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// A delegate to represent the callbacks signatures used in the listener.
+    /// </summary>
+    public delegate void MeasurementCallback<T>(
+                            Instrument instrument,
+                            T measurement,
+                            ReadOnlySpan<KeyValuePair<string, object?>> tags,
+                            object? cookie);
```

Comment:
@tarekgh - do you know if it is common for us to name data like this 'cookie'? I am fine with it as-is, but if there is a better convention such as calling it 'state', 'userData', 'userState', 'userToken', etc. I am glad to change it. For example I see [Timer](https://docs.microsoft.com/en-us/dotnet/api/system.threading.timer.-ctor?view=net-5.0#System_Threading_Timer__ctor_System_Threading_TimerCallback_System_Object_System_Int32_System_Int32_) and [WaitCallback](https://docs.microsoft.com/en-us/dotnet/api/system.threading.waitcallback?view=net-5.0) call their similar parameter 'state'

Reply:
I don't think we have a design guideline to name such parameter more than 

✔️ [DO use descriptive parameter names.](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/naming-parameters)

But trying to look at more cases like: [contextcallback](https://docs.microsoft.com/en-us/dotnet/api/system.threading.contextcallback?view=net-5.0), [SendOrPostCallback](https://docs.microsoft.com/en-us/dotnet/api/system.threading.sendorpostcallback?view=net-5.0), and [WaitOrTimerCallback](https://docs.microsoft.com/en-us/dotnet/api/system.threading.waitortimercallback?view=net-5.0) and I am seeing it is using the name `state`. Considering that, I'll change the name from cookie to state.

Reply:
`state` sounds better.